plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/ProjectUtils.java

```diff
@@ -178,19 +183,43 @@ public void run() {
 
   /** Wait for any spawned jobs and builds to complete (e.g., validation jobs). */
   public static void waitForProjects(Runnable delayTactic, IProject... projects) {
+    IJobManager jobManager = Job.getJobManager();
     try {
+      Set<Job> jobs = new LinkedHashSet<>();
+      List<String> allBuildErrors;
       do {
-        Job.getJobManager().join(ResourcesPlugin.FAMILY_MANUAL_BUILD, null);
-        Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+        delayTactic.run();
+        for (Job job : jobs) {
+          job.join();
+        }
+        jobs.clear();
+
+        Collections.addAll(jobs, jobManager.find(ResourcesPlugin.FAMILY_MANUAL_BUILD));
+        Collections.addAll(jobs, jobManager.find(ResourcesPlugin.FAMILY_AUTO_BUILD));
         // J2EEElementChangedListener.PROJECT_COMPONENT_UPDATE_JOB_FAMILY
-        Job.getJobManager().join("org.eclipse.jst.j2ee.refactor.component", null);
+        Collections.addAll(jobs, jobManager.find("org.eclipse.jst.j2ee.refactor.component"));
         // ServerPlugin.SHUTDOWN_JOB_FAMILY
-        Job.getJobManager().join("org.eclipse.wst.server.core.family", null);
-        Job.getJobManager().join("org.eclipse.wst.server.ui.family", null);
-        ValidationFramework.getDefault().join(null);
-
-        delayTactic.run();
-      } while (!getAllBuildErrors(projects).isEmpty());
+        Collections.addAll(jobs, jobManager.find("org.eclipse.wst.server.core.family"));
+        Collections.addAll(jobs, jobManager.find("org.eclipse.wst.server.ui.family"));
+        Collections.addAll(jobs, jobManager.find(ValidationBuilder.FAMILY_VALIDATION_JOB));
+        allBuildErrors = getAllBuildErrors(projects);
+        if (DEBUG) {
+          if (!jobs.isEmpty()) {
+            System.err.printf("ProjectUtils#waitForProjects: waiting for %d jobs: %s\n",
+                jobs.size(), jobs);
+          } else if (!allBuildErrors.isEmpty()) {
+            System.err.printf("ProjectUtils#waitForProjects: waiting for %d build errors\n",
+                allBuildErrors.size());
+          } else {
+            Job[] otherJobs = jobManager.find(null);
```

Comment:
Fair point though it's only executed if there were no sought-for jobs found, so it really will be the other jobs.